messages/teleporter/message_manager.go

```diff
@@ -90,16 +90,43 @@ func isAllowedRelayer(allowedRelayers []common.Address, eoa common.Address) bool
 	return false
 }
 
-// ShouldSendMessage returns true if the message should be sent to the destination chain
-func (m *messageManager) ShouldSendMessage(warpMessageInfo *vmtypes.WarpMessageInfo, destinationChainID ids.ID) (bool, error) {
-	// Unpack the teleporter message and add it to the cache
+func (m *messageManager) GetDestinationChainID(warpMessageInfo *vmtypes.WarpMessageInfo) (ids.ID, error) {
+	// Unpack the Teleporter message to get the destination chain ID
 	teleporterMessage, err := teleportermessenger.UnpackTeleporterMessage(warpMessageInfo.WarpPayload)
 	if err != nil {
 		m.logger.Error(
 			"Failed unpacking teleporter message.",
 			zap.String("warpMessageID", warpMessageInfo.WarpUnsignedMessage.ID().String()),
 		)
-		return false, err
+		return ids.ID{}, err
+	}
+
+	// Cache the message so it can be reused in SendMessage
+	m.teleporterMessageCache.Put(warpMessageInfo.WarpUnsignedMessage.ID(), teleporterMessage)
+
+	return teleporterMessage.DestinationChainID, nil
+}
+
+// ShouldSendMessage returns true if the message should be sent to the destination chain
+func (m *messageManager) ShouldSendMessage(warpMessageInfo *vmtypes.WarpMessageInfo, destinationChainID ids.ID) (bool, error) {
+	warpMssageID := warpMessageInfo.WarpUnsignedMessage.ID()
+	teleporterMessage, ok := m.teleporterMessageCache.Get(warpMssageID)
+	if !ok {
+		m.logger.Debug(
+			"Teleporter message to send not in cache. Extracting from signed warp message.",
+			zap.String("destinationChainID", destinationChainID.String()),
+			zap.String("warpMessageID", warpMssageID.String()),
+		)
+		var err error
+		teleporterMessage, err = teleportermessenger.UnpackTeleporterMessage(warpMessageInfo.WarpPayload)
+		if err != nil {
+			m.logger.Error(
+				"Failed unpacking teleporter message.",
+				zap.String("destinationChainID", destinationChainID.String()),
+				zap.String("warpMessageID", warpMssageID.String()),
+			)
+			return false, err
+		}
```

Comment:
we should put in cache here

Reply:
Actually we should only put to cache if `cache.Get` didn't return an entry anyways, so no need to put at the end in case we already have cache entry.

Reply:
Think we can also put to cache for `sendMessage` if no cache entry already.

Reply:
Commented above.